File: app/auth/modules/bearer_redis.py
from __future__ import annotations

import os
from typing import Any

import httpx
from redis.asyncio import Redis

from app.auth.base import AuthContext, AuthResult, _invalidate_on_from_config


def _walk_path(body: dict | None, path: str) -> list:
    """Subset of JSONPath: dotted segments, optional [] for arrays. Returns flat list of leaf values."""
    if body is None:
        return []
    nodes = [body]
    for segment in path.split("."):
        next_nodes = []
        if segment.endswith("[]"):
            key = segment[:-2]
            for n in nodes:
                if isinstance(n, dict) and key in n:
                    v = n[key]
                    if isinstance(v, list):
                        next_nodes.extend(v)
        else:
            for n in nodes:
                if isinstance(n, dict) and segment in n:
                    next_nodes.append(n[segment])
        nodes = next_nodes
    return nodes


class BearerRedisAuth:
    """Bearer token stored in Redis; optional env seed and body-error invalidation."""

    name = "bearer_redis"

    def __init__(
        self,
        redis: Redis,
        redis_key: str,
        env_seed: str | None,
        header: str,
        prefix: str,
        invalidate_on: set[int],
        body_json_path: str | None,
        body_codes: set[str],
    ) -> None:
        self._redis = redis
        self._redis_key = redis_key
        self._env_seed = env_seed
        self._header = header
        self._prefix = prefix
        self._invalidate_on = invalidate_on
        self._body_json_path = body_json_path
        self._body_codes = body_codes

    @classmethod
    def from_config(
        cls, config: dict[str, Any], *, redis: Redis, http: httpx.AsyncClient
    ) -> BearerRedisAuth:
        body_codes_cfg = config.get("invalidate_on_body_codes")
        body_json_path: str | None = None
        body_codes: set[str] = set()
        if body_codes_cfg:
            body_json_path = body_codes_cfg.get("json_path")
            body_codes = set(body_codes_cfg.get("codes", []))
        return cls(
            redis=redis,
            redis_key=config["redis_key"],
            env_seed=config.get("env_seed"),
            header=config.get("header", "Authorization"),
            prefix=config.get("prefix", "Bearer "),
            invalidate_on=_invalidate_on_from_config(config),
            body_json_path=body_json_path,
            body_codes=body_codes,
        )

    async def apply(self, ctx: AuthContext) -> AuthResult:
        raw = await self._redis.get(self._redis_key)
        if not raw and self._env_seed:
            seed = os.environ.get(self._env_seed)
            if seed:
                await self._redis.set(self._redis_key, seed)
                raw = seed
        if not raw:
            raise RuntimeError(f"Token not configured at redis_key={self._redis_key!r}.")
        token = raw if isinstance(raw, str) else raw.decode()
        return AuthResult(headers={self._header: f"{self._prefix}{token}"})

    async def invalidate(self) -> None:
        await self._redis.delete(self._redis_key)

    def is_rejection(self, status_code: int, body: dict | None) -> bool:
        if status_code in self._invalidate_on:
            return True
        if self._body_json_path and body is not None:
            leaves = _walk_path(body, self._body_json_path)
            if self._body_codes.intersection(str(v) for v in leaves):
                return True
        return False

File: app/ts/src/auth/modules/bearer_redis.ts
import type { AuthContext, AuthResult } from "../../core/types.ts";
import type { AuthModule, AuthDeps } from "../types.ts";
import { invalidateOnFromConfig } from "../types.ts";

/**
 * Subset of JSONPath: dotted segments, optional [] suffix for array scatter.
 * e.g. "errors[].errorCode" → collect every errorCode from the errors array.
 * Returns a flat list of leaf values.
 */
function walkPath(body: unknown, path: string): unknown[] {
  if (body === null || body === undefined) return [];
  let nodes: unknown[] = [body];
  for (const segment of path.split(".")) {
    const nextNodes: unknown[] = [];
    if (segment.endsWith("[]")) {
      const key = segment.slice(0, -2);
      for (const n of nodes) {
        if (n !== null && typeof n === "object" && !Array.isArray(n)) {
          const record = n as Record<string, unknown>;
          if (key in record) {
            const v = record[key];
            if (Array.isArray(v)) {
              nextNodes.push(...v);
            }
          }
        }
      }
    } else {
      for (const n of nodes) {
        if (n !== null && typeof n === "object" && !Array.isArray(n)) {
          const record = n as Record<string, unknown>;
          if (segment in record) {
            nextNodes.push(record[segment]);
          }
        }
      }
    }
    nodes = nextNodes;
  }
  return nodes;
}

export class BearerRedisAuth implements AuthModule {
  static readonly typeName = "bearer_redis";

  private constructor(
    private readonly deps: AuthDeps,
    private readonly redisKey: string,
    private readonly envSeed: string | null,
    private readonly header: string,
    private readonly prefix: string,
    private readonly invalidateOn: Set<number>,
    private readonly bodyJsonPath: string | null,
    private readonly bodyCodes: Set<string>,
  ) {}

  static fromConfig(
    config: Record<string, unknown>,
    deps: AuthDeps,
  ): BearerRedisAuth {
    const redisKey = config["redis_key"];
    if (typeof redisKey !== "string" || !redisKey) {
      throw new Error(
        "BearerRedisAuth: 'redis_key' must be a non-empty string.",
      );
    }

    const envSeed =
      typeof config["env_seed"] === "string" ? config["env_seed"] : null;
    const header =
      typeof config["header"] === "string" ? config["header"] : "Authorization";
    const prefix =
      typeof config["prefix"] === "string" ? config["prefix"] : "Bearer ";
    const invalidateOn = invalidateOnFromConfig(config, [401]);

    let bodyJsonPath: string | null = null;
    let bodyCodes: Set<string> = new Set();
    const bodyCodesCfg = config["invalidate_on_body_codes"];
    if (
      bodyCodesCfg !== null &&
      bodyCodesCfg !== undefined &&
      typeof bodyCodesCfg === "object" &&
      !Array.isArray(bodyCodesCfg)
    ) {
      const cfg = bodyCodesCfg as Record<string, unknown>;
      if (typeof cfg["json_path"] === "string") {
        bodyJsonPath = cfg["json_path"];
      }
      if (Array.isArray(cfg["codes"])) {
        bodyCodes = new Set(
          (cfg["codes"] as unknown[])
            .filter(
              (v): v is string | number =>
                typeof v === "string" || typeof v === "number",
            )
            .map(String),
        );
      }
    }

    return new BearerRedisAuth(
      deps,
      redisKey,
      envSeed,
      header,
      prefix,
      invalidateOn,
      bodyJsonPath,
      bodyCodes,
    );
  }

  async apply(_ctx: AuthContext): Promise<AuthResult> {
    let raw = await this.deps.tokenStorage.get(this.redisKey);

    if (!raw && this.envSeed) {
      const seed = this.deps.env[this.envSeed];
      if (typeof seed === "string" && seed) {
        // Seed into storage; no TTL — operator owns lifecycle.
        await this.deps.tokenStorage.set(this.redisKey, seed);
        raw = seed;
      }
    }

    if (!raw) {
      throw new Error(
        `BearerRedisAuth: token not configured at redis_key=${JSON.stringify(this.redisKey)}.`,
      );
    }

    return { headers: { [this.header]: `${this.prefix}${raw}` } };
  }

  async invalidate(): Promise<void> {
    await this.deps.tokenStorage.delete(this.redisKey);
  }

  isRejection(statusCode: number, body: unknown): boolean {
    if (this.invalidateOn.has(statusCode)) return true;
    if (this.bodyJsonPath && body !== null && body !== undefined) {
      const leaves = walkPath(body, this.bodyJsonPath);
      for (const leaf of leaves) {
        if (this.bodyCodes.has(String(leaf))) return true;
      }
    }
    return false;
  }
}
